scripts/check-docs/src/main.rs

```diff
@@ -0,0 +1,329 @@
+        .into_iter()
+        .map(|unused_end| anyhow!("Missing anchor start for {unused_end:?}"))
+        .collect::<Vec<_>>();
+
+    let start_only = pairs.into_iter().map(|(begin, _)| begin).collect();
+
+    (start_only, chain!(errors, additional_errors).collect())
+}
+
+fn filter_unused_ends<'a>(ends: &'a [Anchor], pairs: &[(Anchor, Anchor)]) -> Vec<&'a Anchor> {
+    ends.iter()
+        .filter(|end| {
+            let end_used_in_pairs = pairs.iter().any(|(_, used_end)| *end == used_end);
+            !end_used_in_pairs
+        })
+        .collect()
+}
+
+fn check_validity_of_anchor_pair(begin: &Anchor, end: &Anchor) -> Option<anyhow::Error> {
+    if begin.line_no > end.line_no {
+        Some(anyhow!("The end of the anchor appears before the beginning. End anchor: {end:?}. Begin anchor: {begin:?}"))
+    } else {
+        None
+    }
+}
+
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+struct Anchor {
+    line_no: usize,
+    name: String,
+    file: PathBuf,
+}
+
+#[allow(dead_code)]
+enum TestEnum {
+    Anchor(Vec<Anchor>),
+    Include(Vec<Include>),
+    Errors(Vec<Error>),
+}
+
+fn extract_starts_and_ends(
+    text_w_anchors: &str,
+) -> anyhow::Result<(Vec<Anchor>, Vec<Anchor>), Error> {
+    let apply_regex = |regex: Regex| {
+        text_w_anchors
+            .lines()
+            .filter_map(|line| regex.captures(line))
+            .map(|capture| {
+                let file = PathBuf::from(&capture[1]).canonicalize()?;
+                let line_no = &capture[2];
+                let anchor_name = &capture[3];
+
+                Ok(Anchor {
+                    line_no: line_no.parse().unwrap(),
+                    name: anchor_name.to_string(),
+                    file,
+                })
+            })
+            .collect::<Result<Vec<_>, Error>>()
+    };
+
+    let begins = apply_regex(Regex::new(
+        r"^(.+):(\d+):\s*(?:/{2,}|/\*)\s*ANCHOR\s*:\s*([\w_-]+)\s*(?:\*/)?",
+    )?)?;
+    let ends = apply_regex(Regex::new(
+        r"^(.+):(\d+):\s*(?:/{2,}|/\*)\s*ANCHOR_END\s*:\s*([\w_-]+)\s*(?:\*/)?",
+    )?)?;
+
+    Ok((begins, ends))
+}
+
+fn search_for_patterns_in_project(pattern: &str) -> anyhow::Result<String> {
+    let grep_project = std::process::Command::new("grep")
+        .arg("-I")
+        .arg("-H")
+        .arg("-R")
+        .arg("-n")
+        .arg("--exclude-dir=scripts")
+        .arg(pattern)
+        .arg(".")
+        .output()
+        .expect("failed grep command");
+
+    if !grep_project.status.success() {
+        bail!("Failed running grep command for searching {}", pattern);
+    }
+
+    Ok(String::from_utf8(grep_project.stdout)?)
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        extract_starts_and_ends, filter_valid_anchors, parse_includes, validate_includes, TestEnum,
+    };
+    use anyhow::bail;
+
+    fn test_search_for_patterns(pattern: &str, test_file: &str) -> anyhow::Result<String> {
+        let grep = std::process::Command::new("grep")
```

Comment:
This is duplicated code. If you wanted to test the `search_for_patterns` you should call it here.

Reply:
Also, this is testing just that the command ran. 

It should rather be testing that you've got the flag combination correct. That the stdout it spat out contains everything you expect it to.

Reply:
Also, this is much more of an integration test -- running a process and depending on external files. Like spinning up a db and expecting it to have prepopulated data.

It is [convention](https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html) to place such tests inside a separate `tests` directory.

Reply:
> This is duplicated code. If you wanted to test the `search_for_patterns` you should call it here

Didn't want to change `search_for_patterns_in_project` so i made `test_search_for_patterns` with additional parameter test_file. I should maybe change its name. In this way, it checks whether `Anchors` and `Includes` are well imported, connected and whether it reacts to errors that may arise during writing anchors and includes in the docs.

Reply:
> Also, this is much more of an integration test -- running a process and depending on external files. Like spinning up a db and expecting it to have prepopulated data.
> 
> It is [convention](https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html) to place such tests inside a separate `tests` directory.

Maybe i should do that. But I saw that many of our tests are written in the same file as the code, so I did so.